### PR TITLE
[SPARK-15141][EXAMPLE][DOC] Update OneVsRest Examples

### PR DESCRIPTION
**Pull request by zhengruifeng**
## What changes were proposed in this pull request?

1, Add python example for OneVsRest
2, remove args-parsing
## How was this patch tested?

manual tests
`./bin/spark-submit examples/src/main/python/ml/one_vs_rest_example.py`

### CODE REVIEW

**Comment by zhengruifeng** [comment]
There seems no similar API to directly get `numClass` like scala
`val numClasses = MetadataUtils.getNumClasses(predictionColSchema).get`

**Comment by HyukjinKwon** [comment]
I am not sure if fixing examples can have the component  `[DOC]` in the title. I saw `[EXAMPLE]` component was used by @dongjoon-hyun. This is a pretty minor but making good examples of PRs will help all other contributors. 

@dongjoon-hyun Do you mind If I ask your thoughts please?

**Comment by MLnick** [comment]
@HyukjinKwon I get your concern, but these examples are actually directly included in the HTML documentation for Spark, via `include_example`, so they are both examples and docs. For this kind of thing that touches the HTML doc I think `[DOC]` is important to include.

**Comment by MLnick** [comment]
@zhengruifeng I think the `OneVsRest` examples are a little bloated. Could we turn them into simpler versions (essentially just the `run` part of the example) similar to say `GradientBoostedTreeClassifierExample`. Same applies for Java.

**Comment by zhengruifeng** [comment]
@MLnick Agreed. I will remove the args-parsing blocks in the three example files.

**Comment by yanboliang** [comment]
+1 @MLnick 
There are two types of example code under ```examples/```:
* Example applications (the ones with command-line parsing). Those are served as template code for users to build their own standalone applications. We should create those only for important algorithms.
* Tutorial code. Those are for the user guide, usually without options to play with. We should keep one example per algorithm.

File: examples/src/main/python/ml/one_vs_rest_example.py [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    assert params.fitIntercept in ("true", "false")
+    params.fitIntercept = params.fitIntercept == "true"
+
+    return params
+
+if __name__ == "__main__":
+
+    params = parse()
+
+    sc = SparkContext(appName="PythonOneVsRestExample")
```
Comment (yanboliang) [inline examples/src/main/python/ml/one_vs_rest_example.py]: Use ```SparkSession``` instead of ```SQLContext```. See #12809 for details.

Comment (zhengruifeng) [inline examples/src/main/python/ml/one_vs_rest_example.py]: Thanks!

**Comment by zhengruifeng** [comment]
@MLnick Args-Parsing was removed in those examples

File: examples/src/main/scala/org/apache/spark/examples/ml/OneVsRestExample.scala [hunk 2]
```diff
@@ -18,171 +18,71 @@
...
-    val (predictionDuration, predictions) = time(ovrModel.transform(test))
+    val predictions = ovrModel.transform(test)
 
-    // evaluate the model
-    val predictionsAndLabels = predictions.select("prediction", "label")
-      .rdd.map(row => (row.getDouble(0), row.getDouble(1)))
-
-    val metrics = new MulticlassMetrics(predictionsAndLabels)
+    // obtain metrics.
+    val metrics = new MulticlassMetrics(predictions.as[(Double, Double)].rdd)
```
Comment (MLnick) [inline examples/src/main/scala/org/apache/spark/examples/ml/OneVsRestExample.scala]: I think we should prefer to use `ml.evaluation.MulticlassClassificationEvaluator` here as it's a DataFrame API example. You may have to change the metric used - see `DecisionTreeClassificationExample` for example

File: examples/src/main/python/ml/one_vs_rest_example.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+    # train the multiclass model.
+    ovrModel = ovr.fit(train)
+
+    # score the model on test data.
+    predictions = ovrModel.transform(test)
+
+    # obtain metrics.
+    predictionAndLabels = predictions.rdd.map(lambda r: (r.prediction, r.label))
+    metrics = MulticlassMetrics(predictionAndLabels)
```
Comment (MLnick) [inline examples/src/main/python/ml/one_vs_rest_example.py]: same here, use ml evaluator.

File: examples/src/main/java/org/apache/spark/examples/ml/JavaOneVsRestExample.java [hunk 4]
```diff
@@ -17,102 +17,66 @@
...
     OneVsRestModel ovrModel = ovr.fit(train.cache());
 
-    // score the model on test data
-    Dataset<Row> predictions = ovrModel.transform(test.cache())
+    // score the model on test data.
+    Dataset<Row> predictions = ovrModel.transform(test)
       .select("prediction", "label");
 
-    // obtain metrics
+    // obtain metrics.
```
Comment (MLnick) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaOneVsRestExample.java]: same here, use ml evaluator

File: examples/src/main/python/ml/one_vs_rest_example.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+    # $example on$
+    # load data file.
+    inputData = spark.read.format("libsvm") \
+        .load("data/mllib/sample_multiclass_classification_data.txt")
+
+    # generate the train/test split.
+    (train, test) = inputData.randomSplit([0.8, 0.2])
+
+    # instantiate the base classifier.
+    lrParams = {'maxIter': 10, 'tol': 1E-6, 'fitIntercept': True}
```
Comment (MLnick) [inline examples/src/main/python/ml/one_vs_rest_example.py]: This is a little strange, why not simply `LogisticRegression(maxIter=10, tol=...)`?

File: examples/src/main/python/ml/one_vs_rest_example.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+The example uses Logistic Regression as the base classifier.
+Run with:
+  bin/spark-submit examples/src/main/python/ml/one_vs_rest_example.py
+"""
+
+
+if __name__ == "__main__":
+    spark = SparkSession \
+        .builder \
+        .appName("OneHotEncoderExample") \
```
Comment (MLnick) [inline examples/src/main/python/ml/one_vs_rest_example.py]: Incorrect app name here

File: examples/src/main/scala/org/apache/spark/examples/ml/OneVsRestExample.scala [hunk 7]
```diff
@@ -18,171 +18,71 @@
...
-    params.regParam.foreach(classifier.setRegParam)
-    params.elasticNetParam.foreach(classifier.setElasticNetParam)
+      .setMaxIter(10)
+      .setTol(1E-6)
+      .setFitIntercept(true)
 
     // instantiate the One Vs Rest Classifier.
-
     val ovr = new OneVsRest()
     ovr.setClassifier(classifier)
```
Comment (MLnick) [inline examples/src/main/scala/org/apache/spark/examples/ml/OneVsRestExample.scala]: minor but perhaps use `val over = new OneVsRest().setClassifier(...)` to match Java example.

**Comment by MLnick** [comment]
@zhengruifeng made a few comments - most importantly it's better to use the ml evaluators throughout as these are DataFrame API examples.

**Comment by zhengruifeng** [comment]
@MLnick `MulticlassClassificationEvaluator` do not support `confusionMatrix` not. So I will just remove the computaion of `confusionMatrix`.

**Comment by MLnick** [comment]
Just use accuracy I think similar to the decision tree example

File: examples/src/main/java/org/apache/spark/examples/ml/JavaOneVsRestExample.java [hunk 8]
```diff
@@ -17,222 +17,69 @@
...
 
-    Matrix confusionMatrix = metrics.confusionMatrix();
-    // output the Confusion Matrix
-    System.out.println("Confusion Matrix");
-    System.out.println(confusionMatrix);
-    System.out.println();
-    System.out.println(results);
+    // compute the classification error on test data.
+    double precision = evaluator.evaluate(predictions);
+    System.out.print("Test Error : " + (1 - precision));
```
Comment (srowen) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaOneVsRestExample.java]: Nit: println

Comment (zhengruifeng) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaOneVsRestExample.java]: Thanks!

File: examples/src/main/java/org/apache/spark/examples/ml/JavaOneVsRestExample.java [hunk 9]
```diff
@@ -17,222 +17,69 @@
...
-import org.apache.commons.cli.*;
-
 // $example on$
 import org.apache.spark.ml.classification.LogisticRegression;
 import org.apache.spark.ml.classification.OneVsRest;
 import org.apache.spark.ml.classification.OneVsRestModel;
-import org.apache.spark.ml.util.MetadataUtils;
-import org.apache.spark.mllib.evaluation.MulticlassMetrics;
+import org.apache.spark.ml.evaluation.MulticlassClassificationEvaluator;
 import org.apache.spark.mllib.linalg.Matrix;
```
Comment (MLnick) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaOneVsRestExample.java]: This import is no longer required I believe

File: examples/src/main/java/org/apache/spark/examples/ml/JavaOneVsRestExample.java [hunk 10]
```diff
@@ -17,222 +17,69 @@
...
 import org.apache.spark.sql.Dataset;
 import org.apache.spark.sql.Row;
-import org.apache.spark.sql.SparkSession;
-import org.apache.spark.sql.types.StructField;
 // $example off$
+import org.apache.spark.sql.SparkSession;
+
 
 /**
  * An example runner for Multiclass to Binary Reduction with One Vs Rest.
```
Comment (MLnick) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaOneVsRestExample.java]: you can remove "runner" in this doc string (in this example and the others)

**Comment by MLnick** [comment]
@zhengruifeng made a few comments. Pending those I think this is ready

**Comment by zhengruifeng** [comment]
@MLnick Thanks. Updated

File: examples/src/main/java/org/apache/spark/examples/ml/JavaOneVsRestExample.java [hunk 11]
```diff
@@ -17,222 +17,68 @@
...
-      double f = params.fracTest;
-      Dataset<Row>[] tmp = inputData.randomSplit(new double[]{1 - f, f}, 12345);
-      train = tmp[0];
-      test = tmp[1];
-    }
+    // instantiate the One Vs Rest Classifier.
+    OneVsRest ovr = new OneVsRest().setClassifier(classifier);
 
-    // train the multiclass model
+    // train the multiclass model.
```
Comment (MLnick) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaOneVsRestExample.java]: Sorry one last thing - here we use `train.cache()` but we don't do that in the other examples. Actually in general we don't seem to do that in any other examples from a quick look. So perhaps remove that and just do `ovr.fit(train);`

Comment (zhengruifeng) [inline examples/src/main/java/org/apache/spark/examples/ml/JavaOneVsRestExample.java]: Right. I will fix it

**Comment by MLnick** [comment]
LGTM, merged to master and branch-2.0. Thanks!

**Pull request closed without merging**